drivers/ina3221/ina3221.c

```diff
@@ -0,0 +1,928 @@
+}
+
+/**
+ * @brief Convert register value to shunt voltage sum value in mV
+ * 
+ * @param[in]   sum_reg_val Register value
+ * 
+ * @pre         sum_reg_val must be in host byte order
+ * 
+ * @returm      Shunt voltage sum value in mV
+ */
+static inline int32_t sum_reg_val_to_shunt_voltage_uV(int16_t sum_reg_val)
+{
+    //assert(sum_reg_val >= INA3221_MIN_SHUNT_SUM_REG_VAL);
+    assert(sum_reg_val <= INA3221_MAX_SHUNT_SUM_REG_VAL);
+
+    int32_t sum_shunt_voltage_uV;
+    if(sum_reg_val < 0)
+    {
+        sum_shunt_voltage_uV = (~(sum_reg_val) + 1) / 2 * INA3221_SHUNT_VOLTAGE_PRECISION_uV;
+        sum_shunt_voltage_uV *= -1;
+    }
+    else
+    {
+        sum_shunt_voltage_uV =  sum_reg_val / 2 * INA3221_SHUNT_VOLTAGE_PRECISION_uV;
+    }
+    return sum_shunt_voltage_uV;
+}
+
+/**
+ * @brief Convert shunt voltage sum value in uV to register value
+ * 
+ * @param[in]   sum_suV shunt voltage sum value in uV
+ * 
+ * @post        Output register value is in host byte order
+ * 
+ * @return      Register value
+ */
+static inline int16_t sum_shunt_voltage_uV_to_reg_val(int32_t sum_suV)
+{
+    //assert(sum_suV >= INA3221_MIN_SHUNT_SUM_uV);
+    assert(sum_suV <= INA3221_MAX_SHUNT_SUM_uV);
+
+    int16_t reg_val;
+    if(sum_suV < 0)
+    {
+        sum_suV *= (-1);
+        reg_val = ((~(sum_suV / INA3221_SHUNT_VOLTAGE_PRECISION_uV * 2)) + 1);
+    }
+    else
+    {
+        reg_val = sum_suV / INA3221_SHUNT_VOLTAGE_PRECISION_uV * 2;
+    }
+    return reg_val; 
+}
+
+/**
+ * @brief Read register value
+ * 
+ * @param[in]   dev Device handle
+ * @param[in]   reg Register address
+ * @param[out]  out Output register value
+ * 
+ * @post        @p out is in host byte order
+ * 
+ * @return      0, on success
+ * @return      -INA3221_I2C_ERROR, if i2c bus acquirement/release failed
+ * @return      @see i2c_read_regs
+ */
+static int _read_reg(const ina3221_t* dev, uint8_t reg, uint16_t* out)
+{
+    if(i2c_acquire(dev->params.i2c))
+    {
+        return -INA3221_I2C_ERROR;
+    }
+    int status = i2c_read_regs(dev->params.i2c, dev->params.addr, reg, out, INA3221_REG_LEN, 0);
+    if(i2c_release(dev->params.i2c))
+    {
+        return -INA3221_I2C_ERROR;
+    }
+    if(status < 0)
+    {
+        return status;
+    }
+    *out = ntohs(*out);
+    return 0;
+}
+
+/**
+ * @brief Write register value
+ * 
+ * @param[in]   dev Device handle
+ * @param[in]   reg Register address
+ * @param[out]  in Input register value
+ * 
+ * @pre         @p in must be in host byte order
+ * 
+ * @return      0, on success
+ * @return      -INA3221_I2C_ERROR, if i2c bus acquirement/release failed
+ * @retturn     @see i2c_write_regs
+ */
+static int _write_reg(const ina3221_t* dev, uint8_t reg, uint16_t in)
+{
+    in = htons(in);
+    if(i2c_acquire(dev->params.i2c))
+    {
+        return -INA3221_I2C_ERROR;
+    }
+    int status = i2c_write_regs(dev->params.i2c, dev->params.addr, reg, &in, INA3221_REG_LEN, 0);
+    if(i2c_release(dev->params.i2c))
```

Comment:
I just looked up and can confirm that `i2c_release()` indeed can fail. This is quite surprising to me. One would expect that returning a resource should never fail, as there is no reasonable error handling that could be done...

I suggest to just ignore the return value of `i2c_release()`, as really no reasonable error handling could be done.

Reply:
Ok, I will drop the error checking for `i2c_release`. But I personally think `i2c_acquire/i2c_release`
should be like in _cpu/cc2538/periph_. (e.g.),
because it checks user input of public API. An assert would be removed in release mode and the user perhaps would not recognize that the program shows weird behaviour because he configured an invalid I2C bus index, if there are more than 1 mutexes, as e.g. in _cpu/efm32/periph_.
```
int i2c_acquire(i2c_t dev)
{
    DEBUG("%s\n", __FUNCTION__);
    if (dev < I2C_NUMOF) {
        mutex_lock(&lock);
        return 0;
    }
    return -1;
}

int i2c_release(i2c_t dev)
{
    DEBUG("%s\n", __FUNCTION__);
    if (dev < I2C_NUMOF) {
        mutex_unlock(&lock);
        return 0;
    }
    return -1;
}
```

Reply:
I think in the standard use case any `i2c_t`, `spi_t` and the like will be a compile time constant. So if no `assert()`s blew up during development, there is no risk of bad things happening in production builds for this use case. And for the use case `i2c_t` and `spi_t` really are evaluated at run-time (e.g. as in the tests applications where it can be set via shell command), it could be checked once directly after the user supplied the value.

For `i2c_release()` an assert should however be fine in any case: This really can only legitimately fail if a different device was released than acquired. And this is not an error to expect and handle, but a bug to fix. So `assert()` would be the right tool for that case.